wandb/sdk/wandb_settings.py

```diff
@@ -954,6 +960,38 @@ def _validate__stats_samples_to_average(value: int) -> bool:
             raise UsageError("_stats_samples_to_average must be between 1 and 30")
         return True
 
+    @staticmethod
+    def _validate_async_upload_concurrency_limit(value: int) -> bool:
+        if not isinstance(value, int):
+            raise UsageError("async_upload_concurrency_limit must be an int")
+
+        if value <= 0:
+            raise UsageError("async_upload_concurrency_limit must be positive")
+
+        try:
+            import resource  # not always available on Windows
+
+            file_limit = resource.getrlimit(resource.RLIMIT_NOFILE)[0]
+        except Exception:
+            # Couldn't get the open-file-limit for some reason,
+            # probably very platform-specific. Not a problem,
+            # we just won't use it to cap the concurrency.
+            pass
+        else:
+            if value > file_limit:
+                wandb.termwarn(
+                    (
+                        "async_upload_concurrency_limit setting of"
+                        + f"{value} exceeds this process's limit"
+                        + f" on open files ({file_limit}); may cause file-upload failures."
+                        + " Try decreasing async_upload_concurrency_limit,"
+                        + " or increasing your file limit with `ulimit -n`."
```

Comment:
Automatic string concatenation means you don't need the `+`s here:
```suggestion
                        "async_upload_concurrency_limit setting of"
                        f" {value} exceeds this process's limit"
                        f" on open files ({file_limit}); may cause file-upload failures."
                        " Try decreasing async_upload_concurrency_limit,"
                        " or increasing your file limit with `ulimit -n`."
```

Reply:
...yeah, sure, all right. I'm twice shy because I've been burned by string concatenation before, but I think here I'm just being silly.

(Storytime for the interested: at a previous job, we had an important list of string constants
```python
things = (
    "foo",
    "bar",
    # ...
    "baz"
)
```
...and occasionally somebody would add an entry...
```python
things = (
    "foo",
    "bar",
    # ...
    "baz"
    "quux"
)
```
...and it would cause all kinds of trouble.

I guess we must not have had any kind of self-respecting linter or code-format-enforcer? I forget.)